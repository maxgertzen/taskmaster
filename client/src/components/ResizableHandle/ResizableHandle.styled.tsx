import styled from '@emotion/styled';

export const StyledResizableHandle = styled.div(({ theme }) => ({
  position: 'absolute',
  top: 0,
  right: 0,
  width: theme.spacing(0.5),
  height: '100%',
  cursor: 'col-resize',
  backgroundColor: 'transparent',
}));
