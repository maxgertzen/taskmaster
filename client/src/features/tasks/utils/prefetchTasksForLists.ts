import { QueryClient } from '@tanstack/react-query';

import { QUERY_KEYS } from '@/shared/api/query-keys';
import { STALE_TIME } from '@/shared/constants/staleTime';

import { fetchTasks } from '../api';

export const prefetchTasksForLists = (
  listIds: string[],
  queryClient: QueryClient,
  token: string | null
) => {
  if (!token || !listIds.length) return;

  listIds.forEach((listId) => {
    queryClient.prefetchQuery({
      queryKey: QUERY_KEYS.tasks({ listId }),
      queryFn: fetchTasks(token),
      staleTime: STALE_TIME,
    });
  });
};
