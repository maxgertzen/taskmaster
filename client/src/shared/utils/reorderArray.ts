export const reorderArray = <T>(
  array: T[],
  fromIndex: number,
  toIndex: number
): T[] => {
  const reordered = [...array];
  const [movedItem] = reordered.splice(fromIndex, 1);
  reordered.splice(toIndex, 0, movedItem);
  return reordered;
};
