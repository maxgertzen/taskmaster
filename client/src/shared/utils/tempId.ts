export const generateTempId = (prefix: string = 'temp') =>
  `${prefix}-${Date.now()}`;
