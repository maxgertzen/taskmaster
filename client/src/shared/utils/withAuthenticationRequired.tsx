import { withAuthenticationRequired as realWithAuthenticationRequired } from '@auth0/auth0-react';

import { IS_AUTH0_DISABLED } from '@/constants';
import { mockWithAuthenticationRequired } from '@/mocks';

const isMock = IS_AUTH0_DISABLED === 'true';

export const withAuthenticationRequired = isMock
  ? mockWithAuthenticationRequired
  : realWithAuthenticationRequired;
