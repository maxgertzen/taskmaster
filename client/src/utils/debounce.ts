export function debounce<T extends (...args: Parameters<T>) => ReturnType<T>>(
  func: T,
  delay: number
): ((...args: Parameters<T>) => void) & { cancel: () => void } {
  let timeoutId: ReturnType<typeof setTimeout> | null = null;

  const debouncedFunction = (...args: Parameters<T>) => {
    if (timeoutId) clearTimeout(timeoutId);
    timeoutId = setTimeout(() => func(...args), delay);
  };

  debouncedFunction.cancel = () => {
    if (timeoutId) clearTimeout(timeoutId);
  };

  return debouncedFunction;
}
