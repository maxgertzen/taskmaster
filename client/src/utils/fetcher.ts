export type FetcherConfig<K> = {
  url: string;
  method?: 'GET' | 'POST' | 'PUT' | 'DELETE';
  body?: K;
  timeout?: number;
  headers?: HeadersInit;
  urlSearchParams?: string;
};

export const fetcher = async <T, K = undefined>(
  config: FetcherConfig<K>
): Promise<T> => {
  try {
    const {
      method = 'GET',
      body = {},
      headers = {},
      timeout = 5000,
      urlSearchParams,
    } = config;
    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), timeout);

    const options: RequestInit = {
      method,
      headers: {
        'Content-Type': 'application/json',
        ...headers,
      },
      ...(method !== 'GET' && { body: JSON.stringify(body) }),
      signal: controller.signal,
    };

    if (urlSearchParams) {
      config.url += `?${urlSearchParams}`;
    }

    const response = await fetch(config.url, options);
    clearTimeout(timeoutId);

    if (!response.ok) {
      throw new Error(`Error: ${response.status} ${response.statusText}`);
    }

    return await response.json();
  } catch (error) {
    console.error('Failed to fetch data:', error);
    throw error;
  }
};
