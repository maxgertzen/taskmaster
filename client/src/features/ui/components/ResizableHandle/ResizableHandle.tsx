import React, { FC } from 'react';

import { StyledResizableHandle } from './ResizableHandle.styled';

interface ResizableHandleProps {
  onResize: (isResizing: boolean) => void;
  minWidth?: number;
  maxWidth?: number;
  initialWidth: number;
  setWidth: (newWidth: number) => void;
}

export const ResizableHandle: FC<ResizableHandleProps> = ({
  onResize,
  minWidth = 150,
  maxWidth = 500,
  initialWidth,
  setWidth,
}) => {
  const handleMouseDown = (e: React.MouseEvent<HTMLDivElement>) => {
    const startX = e.clientX;

    const onMouseMove = (event: MouseEvent) => {
      const deltaX = event.clientX - startX;
      const newWidth = initialWidth + deltaX;
      const clampedWidth = Math.max(minWidth, Math.min(maxWidth, newWidth));

      setWidth(clampedWidth);
      onResize(true);
    };

    const onMouseUp = () => {
      document.removeEventListener('mousemove', onMouseMove);
      document.removeEventListener('mouseup', onMouseUp);
      onResize(false);
    };

    document.addEventListener('mousemove', onMouseMove);
    document.addEventListener('mouseup', onMouseUp);
  };

  return <StyledResizableHandle onMouseDown={handleMouseDown} />;
};
