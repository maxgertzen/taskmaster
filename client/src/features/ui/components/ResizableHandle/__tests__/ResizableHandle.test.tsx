import { fireEvent } from '@testing-library/dom';
import { act } from '@testing-library/react';
import { describe, it, expect, vi, beforeEach } from 'vitest';

import { render } from '@/tests/utils';

import { ResizableHandle } from '../ResizableHandle';

describe('ResizableHandle', () => {
  const mockOnResize = vi.fn();
  const mockSetWidth = vi.fn();
  const initialProps = {
    onResize: mockOnResize,
    initialWidth: 200,
    setWidth: mockSetWidth,
  };

  beforeEach(() => {
    vi.clearAllMocks();
  });

  it.each([
    ['light', undefined],
    ['dark', 'dark'],
  ])('renders correctly in %s theme', (_, theme) => {
    const { container } = render(<ResizableHandle {...initialProps} />, {
      theme: theme as 'light' | 'dark',
    });
    expect(container).toMatchSnapshot();
  });

  describe('Resize Behavior', () => {
    it('handles mouse drag within bounds', async () => {
      const { container } = render(<ResizableHandle {...initialProps} />);
      const handle = container.firstChild as HTMLElement;

      await act(async () => {
        // Start resize
        fireEvent.mouseDown(handle, { clientX: 200 });

        // Drag
        fireEvent.mouseMove(document, { clientX: 250 });
        expect(mockSetWidth).toHaveBeenCalledWith(250);
        expect(mockOnResize).toHaveBeenCalledWith(true);

        // End resize
        fireEvent.mouseUp(document);
        expect(mockOnResize).toHaveBeenCalledWith(false);
      });
    });

    it('respects minimum width constraint', async () => {
      const { container } = render(
        <ResizableHandle {...initialProps} minWidth={150} />
      );
      const handle = container.firstChild as HTMLElement;

      await act(async () => {
        fireEvent.mouseDown(handle, { clientX: 200 });
        fireEvent.mouseMove(document, { clientX: 100 }); // Try to drag below min
      });

      expect(mockSetWidth).toHaveBeenCalledWith(150);
    });

    it('respects maximum width constraint', async () => {
      const { container } = render(
        <ResizableHandle {...initialProps} maxWidth={500} />
      );
      const handle = container.firstChild as HTMLElement;

      await act(async () => {
        fireEvent.mouseDown(handle, { clientX: 200 });
        fireEvent.mouseMove(document, { clientX: 800 }); // Try to drag above max
      });

      expect(mockSetWidth).toHaveBeenCalledWith(500);
    });

    it('cleans up event listeners on mouse up', async () => {
      const { container } = render(<ResizableHandle {...initialProps} />);
      const handle = container.firstChild as HTMLElement;

      await act(async () => {
        fireEvent.mouseDown(handle, { clientX: 200 });
        fireEvent.mouseUp(document);

        // Additional mouse moves should not trigger width changes
        fireEvent.mouseMove(document, { clientX: 300 });
      });

      expect(mockSetWidth).not.toHaveBeenCalledWith(300);
    });
  });
});
