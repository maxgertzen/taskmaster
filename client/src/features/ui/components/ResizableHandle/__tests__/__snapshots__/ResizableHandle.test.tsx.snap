// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ResizableHandle > renders correctly in dark theme 1`] = `
<div>
  <div
    class="css-omm4x9"
  />
</div>
`;

exports[`ResizableHandle > renders correctly in light theme 1`] = `
<div>
  <div
    class="css-omm4x9"
  />
</div>
`;
